packages/ast-spec/src/declaration/VariableDeclaration/spec.ts

```diff
@@ -36,18 +40,65 @@ export interface LetOrConstOrVarDeclaration extends BaseNode {
   kind: 'const' | 'let' | 'var';
 }
 
-export interface UsingInNormalContextDeclaration extends BaseNode {
-  type: AST_NODE_TYPES.VariableDeclaration;
+export interface LetOrVarDeclaredDeclaration
+  extends LetOrConstOrVarDeclarationBase {
+  kind: 'let' | 'var';
+  declare: true;
   /**
-   * The variables declared by this declaration.
-   * Note that there may be 0 declarations (i.e. `const;`).
+   * In a `declare let` declaration, the declarators must not have definite assignment
+   * assertions or initializers.
    * ```
-   * using x = 1;
-   * using y =1, z = 2;
+   * declare let x: number;
+   * declare let y, z;
    * ```
    */
-  // TODO(#1852) - this should be guaranteed to have at least 1 element in it.
-  declarations: UsingInNomalConextDeclarator[];
+  declarations: DeclareLetOrVarDeclarator[];
+}
+
+export interface LetOrVarNonDeclaredDeclaration
+  extends LetOrConstOrVarDeclarationBase {
+  kind: 'let' | 'var';
+  declare: false;
+  /**
+   * In a `let`/`var` declaration, the declarators may have definite assignment
+   * assertions or initializers, but not both.
+   */
+  declarations: (
+    | LetOrVarDeclaratorDefiniteAssignment
+    | LetOrConstOrVarDeclaratorNoDefiniteAssignment
+  )[];
+}
+
+export interface ConstDeclaredDeclaration
+  extends LetOrConstOrVarDeclarationBase {
+  kind: 'const';
+  declare: true;
+  /**
+   * In a `declare const` declaration, the declarators may have initializers, but
+   * not definite assignment assertions. Each declarator cannot have both an
+   * initializer and a type annotation.
+   */
+  declarations: LetOrConstOrVarDeclaratorNoDefiniteAssignment[];
+}
+
+export interface ConstNonDeclaredDeclaration
+  extends LetOrConstOrVarDeclarationBase {
+  kind: 'const';
+  declare: false;
+  /**
+   * In a `const` declaration, the declarators must have initializers.
+   */
+  declarations: ConstDeclaratorWithInit[];
+}
+
+export type LetOrConstOrVarDeclaration =
+  | LetOrVarDeclaredDeclaration
+  | LetOrVarNonDeclaredDeclaration
+  | ConstDeclaredDeclaration
+  | ConstNonDeclaredDeclaration;
```

Comment:
The biggest change: made `LetOrConstOrVarDeclaration` much stricter.

Reply:
The thing that needs to be checked with these changes is like `const x;` - do we have a syntax error or do we produce an AST.

If it's the latter then we can't refine the types as the types have to reflect the AST we output.

For the specific example of `const x:` - we produce an AST so the types must allow the case of `const` with an `init: null`.

[playground](https://typescript-eslint.io/play/#ts=5.4.5&showAST=es&fileType=.tsx&code=MYewdgzgLgBAHgbgFBA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)